services/src/pro/handlers/datasets.rs

```diff
@@ -0,0 +1,426 @@
+use actix_web::{web, FromRequest};
+
+use crate::{
+    api::model::{datatypes::DatasetId, services::CreateSystemDataset},
+    contexts::AdminSession,
+    datasets::{storage::DatasetStore, upload::Volume},
+    error::{self, Result},
+    handlers::datasets::{
+        adjust_meta_data_path, auto_create_dataset_handler, create_user_dataset_handler,
+        get_dataset_handler, list_datasets_handler, list_volumes_handler,
+        suggest_meta_data_handler,
+    },
+    pro::{
+        contexts::ProContext,
+        datasets::{DatasetPermission, Permission, Role, UpdateDatasetPermissions},
+    },
+    util::{
+        config::{get_config_element, Data},
+        user_input::UserInput,
+        IdResponse,
+    },
+};
+
+pub(crate) fn init_dataset_routes<C>(cfg: &mut web::ServiceConfig)
+where
+    C: ProContext,
+    C::Session: FromRequest,
+    C::DatasetDB: UpdateDatasetPermissions,
+{
+    cfg.service(
+        web::scope("/dataset")
+            .service(web::resource("/suggest").route(web::get().to(suggest_meta_data_handler::<C>)))
+            .service(web::resource("/auto").route(web::post().to(auto_create_dataset_handler::<C>)))
+            .service(web::resource("/volumes").route(web::get().to(list_volumes_handler)))
+            .service(
+                web::resource("/public").route(web::post().to(create_system_dataset_handler::<C>)),
+            )
+            .service(web::resource("/{dataset}").route(web::get().to(get_dataset_handler::<C>)))
+            .service(web::resource("").route(web::post().to(create_user_dataset_handler::<C>))), // must come last to not match other routes
+    )
+    .service(web::resource("/datasets").route(web::get().to(list_datasets_handler::<C>)));
+}
```

Comment:
Why do you define this twice?

Reply:
because I need a different handler for public datasets in the pro version that explicitly sets the permissions. The free handler doesn't know about permissions.